### PR TITLE
Implement ContentDialog.Closing and ContentDialog.Closed

### LINKED ISSUE #1495: Wasm: ContentDialog Closing support
## What would you like to be added:

The Closing event

## Why is this needed:
To get parity with UWP and allow scenarios where the closing should be prevented

## For which Platform:

- [ ] iOS
- [ ] Android
- [x] WebAssembly
- [ ] Windows

## Anything else we need to know?

See https://github.com/yowl/UnoStandardRefTest2/tree/contentdialog-closing for an example

### PR DESCRIPTION
**Pull request by davidjohnoliver**
GitHub Issue (If applicable): fixes #1495

## PR Type

What kind of change does this PR introduce?

- Feature

## What is the current behavior?

`ContentDialog.Closing` and `ContentDialog.Closed` events are not raised

## What is the new behavior?

They're raised, and corresponding event args are correctly supported.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Tested code with current [supported SDKs](../README.md#supported)
- [ ] Docs have been added/updated which fit [documentation template](https://github.com/nventive/Uno/blob/master/doc/.feature-template.md). (for bug fixes / features)
- [x] [Unit Tests and/or UI Tests](doc/articles/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] [Wasm UI Tests](doc/articles/working-with-the-samples-apps.md#running-the-webassembly-ui-tests-snapshots) are not showing unexpected any differences. Validate PR `Screenshots Compare Test Run` results.
- [x] Contains **NO** breaking changes
- [x] Updated the [Release Notes](https://github.com/nventive/Uno/tree/master/doc/ReleaseNotes)
- [x] Associated with an issue (GitHub or internal)

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/Uno.UI/UI/Xaml/Controls/ContentDialog/DeferralManager.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+		private int _deferralsCount;
+
+		public DeferralManager(Action completeHandler)
+		{
+			this._completeHandler = completeHandler;
+		}
+
+		public T GetDeferral()
+		{
+			_deferralsCount++;
```
Comment (dr1rrb) [inline src/Uno.UI/UI/Xaml/Controls/ContentDialog/DeferralManager.cs]: What if we try to `GetDefferal` after a previous one was completed? (so the `_completeHandler()` will already have been invoked)

File: src/Uno.UI/UI/Xaml/Controls/ContentDialog/DeferralManager.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+				if (isCompleted)
+				{
+					throw new InvalidOperationException("Deferral already completed.");
+				}
+
+				isCompleted = true;
+				_deferralsCount--;
+				if (_deferralsCount <= 0)
+				{
+					_completeHandler();
```
Comment (dr1rrb) [inline src/Uno.UI/UI/Xaml/Controls/ContentDialog/DeferralManager.cs]: I would assume that its legal to complete a deferral on a background thread ... so we should make sure to go back on UI thead for this

File: src/Uno.UI/UI/Xaml/Controls/ContentDialog/ContentDialog.cs [hunk 3]
```diff
@@ -63,9 +63,26 @@ private void OnPopupKeyDown(object sender, KeyRoutedEventArgs e)
...
+					_popup.IsOpen = false;
+
+					Closed?.Invoke(this, new ContentDialogClosedEventArgs(result));
+				}
+			}
+			var closingArgs = new ContentDialogClosingEventArgs(Complete);
+
+			Closing?.Invoke(this, closingArgs);
+
+			if (!closingArgs.IsDeferred)
```
[the remark this answers is not in the document]
Comment (davidjohnoliver) [reply src/Uno.UI/UI/Xaml/Controls/ContentDialog/ContentDialog.cs]: Cancel is checked inside the `Complete()` function

[the remark this answers is not in the document]
Comment (dr1rrb) [reply src/Uno.UI/UI/Xaml/Controls/ContentDialog/ContentDialog.cs]: Indeed, I thought we were checking it only after the `Closed`, but it's the opposite we check it after the `Closing` ... so we cannot cancel the `Closed` event, but I would agree that makes lot of sense :)

File: src/Uno.UI/UI/Xaml/Controls/ContentDialog/DeferralManager.cs (hunk 2 above)
Comment (jeromelaban) [reply src/Uno.UI/UI/Xaml/Controls/ContentDialog/DeferralManager.cs]: Most of those APIs are UI Thread only, and we don't validate thread affinity yet.

Comment (dr1rrb) [reply src/Uno.UI/UI/Xaml/Controls/ContentDialog/DeferralManager.cs]: Yeah @davidjohnoliver checked it and it appears that we get an exception when completing from a BG thread ... that's sad.

Comment (davidjohnoliver) [reply src/Uno.UI/UI/Xaml/Controls/ContentDialog/DeferralManager.cs]: Yep. I've added a check.

File: src/Uno.UI/UI/Xaml/Controls/ContentDialog/DeferralManager.cs (hunk 1 above)
Comment (davidjohnoliver) [reply src/Uno.UI/UI/Xaml/Controls/ContentDialog/DeferralManager.cs]: ooh good point

Comment (jeromelaban) [reply src/Uno.UI/UI/Xaml/Controls/ContentDialog/DeferralManager.cs]: What does WinUI do ?

Comment (davidjohnoliver) [reply src/Uno.UI/UI/Xaml/Controls/ContentDialog/DeferralManager.cs]: So for this case:
```csharp
			dialog.Closing += (o, e) =>
			{
				deferral1 = e.GetDeferral();
				deferral1.Complete();
			};
			dialog.Closing += (o, e) =>
			{
				deferral2 = e.GetDeferral();
			};
```

It won't close the dialog until `deferral2.Complete()` is eventually called. Making the change now.

Comment (davidjohnoliver) [reply src/Uno.UI/UI/Xaml/Controls/ContentDialog/DeferralManager.cs]: fixed

**Comment by jeromelaban** [comment]
The UI test [failed on android](https://dev.azure.com/uno-platform/Uno%20Platform/_build/results?buildId=7465&view=ms.vss-test-web.build-test-results-tab&runId=1010550&resultId=100079&paneView=debug)

**Comment by davidjohnoliver** [comment]
> The UI test [failed on android](https://dev.azure.com/uno-platform/Uno%20Platform/_build/results?buildId=7465&view=ms.vss-test-web.build-test-results-tab&runId=1010550&resultId=100079&paneView=debug)

Yeah, I have a feeling on Android it fails to find views inside of popups. On the other hand, the other ContentDialog tests are apparently working...

**Comment by davidjohnoliver** [comment]
> > The UI test [failed on android](https://dev.azure.com/uno-platform/Uno%20Platform/_build/results?buildId=7465&view=ms.vss-test-web.build-test-results-tab&runId=1010550&resultId=100079&paneView=debug)
> 
> Yeah, I have a feeling on Android it fails to find views inside of popups. On the other hand, the other ContentDialog tests are apparently working...

Yeah it wasn't that, fake news. It's actually that I was setting the `Name` programmatically, and the `ContentDescription` is only applied when `Name` is set in Xaml.

I fixed that, UI tests should work now when `Name` is set programmatically.

File: src/Uno.UI/UI/Xaml/FrameworkElement.cs [hunk 4]
```diff
@@ -527,6 +528,13 @@ async void ApplyPhase()
 			}
 		}
 #endif
+		/// <summary>
+		/// When true (normally because the IsUiAutomationMappingEnabled build property is set), setting the <see cref="Name"/> property
+		/// programmatically will also set the appropriate test identifier for Android/iOS. Disabled by default because it may interfere
+		/// with accessibility features in non-testing scenarios.
+		/// </summary>
+		[EditorBrowsable(EditorBrowsableState.Never)]
+		public static bool IsUiAutomationMappingEnabled { get; set; } = false;
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/FrameworkElement.cs]: Can you move that API here: https://github.com/unoplatform/uno/blob/8dcc6f2d3aa5d57fd3ee96efe5fe73855a45a0da/src/Uno.UI/UI/Xaml/FrameworkElementHelper.cs

You can keep the property here, but internal. It's going to help for compatibility management.

Comment (davidjohnoliver) [reply src/Uno.UI/UI/Xaml/FrameworkElement.cs]: done

**Comment by yowl** [comment]
Hi, can we reopen this, please?

The following code in wasm
```
                contentDialog.Closing += (sender, args) =>
                {
                    Console.WriteLine("Closing " + args.Result);
```
Prints None, when you should be able to check for `ContentDialogResult.Primary` as per UWP

**Pull request merged**